website/project/model.py

```diff
@@ -155,19 +155,119 @@ class MetaData(GuidStoredObject):
     date_modified = fields.DateTimeField(auto_now=datetime.datetime.utcnow)
 
 
-def validate_comment_reports(value, *args, **kwargs):
+def validate_reports(value, *args, **kwargs):
     for key, val in value.iteritems():
         if not User.load(key):
             raise ValidationValueError('Keys must be user IDs')
         if not isinstance(val, dict):
             raise ValidationTypeError('Values must be dictionaries')
-        if 'category' not in val or 'text' not in val:
+        if (
+            'category' not in val or
+            'text' not in val or
+            'date' not in val or
+            'retracted' not in val
+        ):
             raise ValidationValueError(
-                'Values must include `category` and `text` keys'
+                ('Values must include `date`, `category`, ',
+                 '`text`, `retracted` keys')
             )
 
 
-class Comment(GuidStoredObject):
+class SpamMixin(StoredObject):
+    """Mixin to add to objects that can be marked as spam.
+    """
+
+    _meta = {
+        'abstract': True
+    }
+
+    UNKNOWN = 0
+    FLAGGED = 1
+    SPAM = 2
+    HAM = 4
+
+    spam_status = fields.IntegerField(default=UNKNOWN, index=True)
+
+    # Reports is a list of reports
+    # Each report is a dictionary including:
+    #  - user: Reporting user
+    #  - date: date reported
+    #  - retracted: if a report has been retracted
+    #  - category: What type of spam does the reporter believe this is
+    #  - message: Comment on the comment
+    reports = fields.DictionaryField(
+        default={}, validate=validate_reports
+    )
+
+    def flag_spam(self, save=False):
+        # If ham and unedited then tell user that they should read it again
+        if self.spam_status == self.UNKNOWN:
+            self.spam_status = self.FLAGGED
+        if save:
+            self.save()
+
+    def remove_flag(self, save=False):
+        if self.spam_status != self.FLAGGED:
+            return
+        for user, report in self.reports.iteritems():
+            if not report.get('retracted', True):
+                return
+        self.spam_status = self.UNKNOWN
+        if save:
+            self.save()
+
+    def confirm_ham(self, save=False):
+        self.spam_status = self.HAM
+        if save:
+            self.save()
+
+    def confirm_spam(self, save=False):
+        self.spam_status = self.SPAM
+        if save:
+            self.save()
+
+    @property
+    def is_spam(self):
+        return self.spam_status == self.SPAM
+
+    def report_spam(self, user, save=False, **kwargs):
```

Comment:
Since this method is used for all types of reports, I think the old name, `report_abuse`, is preferable.